contracts/cw1155-base/src/contract.rs

```diff
@@ -0,0 +1,1239 @@
+    if sender != MINTER.load(deps.storage)? {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut rsp = Response::default();
+
+    let event = execute_transfer_inner(&mut deps, None, Some(&to), &token_id, amount)?;
+    event.add_attributes(&mut rsp);
+
+    rsp.messages = if let Some(msg) = msg {
+        vec![Cw1155ReceiveMsg {
+            operator: info.sender,
+            from: None,
+            amount,
+            token_id: token_id.clone(),
+            msg,
+        }
+        .into_cosmos_msg(to)?]
+    } else {
+        vec![]
+    };
+
+    // insert if not exist
+    let key = TOKENS.key(&token_id);
+    if deps.storage.get(&key).is_none() {
+        key.save(deps.storage, &"".to_owned())?;
+    }
+
+    Ok(rsp)
+}
+
+pub fn execute_burn(
+    env: ExecuteEnv,
+    from: HumanAddr,
+    token_id: TokenId,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    let ExecuteEnv {
+        mut deps,
+        info,
+        env,
+    } = env;
+
+    // whoever can transfer these tokens can burn
+    guard(
+        check_can_approve(deps.as_ref(), &env, &from, &info.sender)?,
+        ContractError::Unauthorized {},
+    )?;
+
+    let mut rsp = Response::default();
+    let event = execute_transfer_inner(&mut deps, Some(&from), None, &token_id, amount)?;
+    event.add_attributes(&mut rsp);
+    Ok(rsp)
+}
+
+pub fn execute_batch_send_from(
+    env: ExecuteEnv,
+    from: HumanAddr,
+    to: HumanAddr,
+    batch: Vec<(TokenId, Uint128)>,
+    msg: Option<Binary>,
+) -> Result<Response, ContractError> {
+    let ExecuteEnv {
+        mut deps,
+        env,
+        info,
+    } = env;
+
+    guard(
+        check_can_approve(deps.as_ref(), &env, &from, &info.sender)?,
+        ContractError::Unauthorized {},
+    )?;
+
+    let mut rsp = Response::default();
+    for (token_id, amount) in batch.iter() {
+        let event = execute_transfer_inner(&mut deps, Some(&from), Some(&to), token_id, *amount)?;
+        event.add_attributes(&mut rsp);
+    }
+
+    rsp.messages = if let Some(msg) = msg {
+        vec![Cw1155BatchReceiveMsg {
+            operator: info.sender,
+            from: Some(from),
+            batch,
+            msg,
+        }
+        .into_cosmos_msg(to)?]
+    } else {
+        vec![]
+    };
+
+    Ok(rsp)
+}
+
+pub fn execute_batch_mint(
+    env: ExecuteEnv,
+    to: HumanAddr,
+    batch: Vec<(TokenId, Uint128)>,
+    msg: Option<Binary>,
+) -> Result<Response, ContractError> {
+    let ExecuteEnv { mut deps, info, .. } = env;
+    let sender = deps.api.canonical_address(&info.sender)?;
+    if sender != MINTER.load(deps.storage)? {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut rsp = Response::default();
+
+    for (token_id, amount) in batch.iter() {
+        let event = execute_transfer_inner(&mut deps, None, Some(&to), &token_id, *amount)?;
+        event.add_attributes(&mut rsp);
+    }
+
+    for (token_id, _) in batch.iter() {
```

Comment:
You don't want to roll this in the same above for loop?

Reply:
> You don't want to roll this in the same above for loop?

thanks, rolled in now, looks better.